superset/common/query_context.py

```diff
@@ -101,21 +104,143 @@ def __init__(  # pylint: disable=too-many-arguments
             "result_format": self.result_format,
         }
 
-    def get_query_result(self, query_object: QueryObject) -> Dict[str, Any]:
-        """Returns a pandas dataframe based on the query object"""
+    @staticmethod
+    def left_join_on_dttm(
+        left_df: pd.DataFrame, right_df: pd.DataFrame
+    ) -> pd.DataFrame:
+        df = left_df.set_index(DTTM_ALIAS).join(right_df.set_index(DTTM_ALIAS))
+        df.reset_index(level=0, inplace=True)
+        return df
+
+    def processing_time_offsets(
+        self, df: pd.DataFrame, query_object: QueryObject,
+    ) -> Tuple[pd.DataFrame, List[str], List[Optional[str]]]:
+        # ensure query_object is immutable
+        query_object_clone = copy.copy(query_object)
+        rv_sql = []
+        cache_keys = []
+
+        time_offsets = query_object.time_offsets
+        outer_from_dttm = query_object.from_dttm
+        outer_to_dttm = query_object.to_dttm
+        for offset in time_offsets:
```

Comment:
I'm not sure you need to run and cache a completely new query for each offset.

Can we somehow compute the final time periods and generate proper `WHERE` conditions with `or` filters instead?

```python
def get_time_periods_for_offsets(time_range, offsets):
    [start, end] = time_range
    periods = [time_range]
    for offset in periods:
        periods.append([start += offset, end += offset])
    return periods
```

Then change https://github.com/apache/superset/blob/bee386e643a202f182a5cffd03c460fc27efe959/superset/connectors/sqla/models.py#L1370-L1375

to something like

```python
inner_time_filter = or_([dttm_col.between(start, end) for start, end in periods])
subq = subq.where(and_(*(where_clause_and + [inner_time_filter]))
```

Reply:
For where clause combined by `or` operator, I estimate that the system consumption is approximately equal to multiple queries. This is because the `or` operator does not reduce rows scan for the database engine. And we don't have the opportunity to cache time offset. Let me explain.

### Use `or` operator in the where clause
- unable to cache every timeoffset slice
- unable to easy generate finial dataframe

![image](https://user-images.githubusercontent.com/2016594/125262268-9e629f00-e334-11eb-97c3-4cafd11d140e.png)



### Use extra query
<img width="831" alt="image" src="https://user-images.githubusercontent.com/2016594/125260485-f26c8400-e332-11eb-991d-5de621ac7aaf.png">

Reply:
＞ This is because the or operator does not reduce rows scan for the database engine. 

It would reduce rows scanned if there are significant overlaps among the offset time periods. E.g. you query for two years of data and offset by 1 year.

> unable to easily generate the final dataframe

Isn't each sub-dataframe a between(start_time, end_time) filter on the query result dataframe? We should probably use pandas to handle the time periods and join by datetime index anyway, if we are not already doing that, so the split & join by time process shouldn't be that hard, either.

Reply:
> ＞ This is because the or operator does not reduce rows scan for the database engine.
> 
> It would reduce rows scanned if there are significant overlaps among the offset time periods. E.g. you query for two years of data and offset by 1 year.
> 
> > unable to easily generate the final dataframe
> 
> Isn't each sub-dataframe a between(start_time, end_time) filter on the query result dataframe? We should probably use pandas to handle the time periods and join by datetime index anyway, if we are not already doing that, so the split & join by time process shouldn't be that hard, either.

I believe moving this type of logic into Superset would be a slippery slope to introducing logic that's usually best handled by the analytical database and cause major maintainability overhead. I'm open to considering this down the road, but it would require some more discussion to ensure we don't end up building a pseudo-database engine inside Superset. Maybe we can revisit this if/when we start working on adding the semantic layer for table joins?